server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -655,17 +654,36 @@ assert node(failedShard.currentNodeId()).getByShardId(failedShard.shardId()) ==
     }
 
     private void unassignPrimaryAndPromoteActiveReplicaIfExists(
-        ShardRouting failedShard,
+        ShardRouting failedPrimary,
         UnassignedInfo unassignedInfo,
         RoutingChangesObserver routingChangesObserver
     ) {
-        assert failedShard.primary();
-        ShardRouting activeReplica = activeReplicaWithHighestVersion(failedShard.shardId());
-        if (activeReplica == null) {
-            moveToUnassigned(failedShard, unassignedInfo);
+        assert failedPrimary.primary();
+        ShardRouting replicaToPromote = activePromotableReplicaWithHighestVersion(failedPrimary.shardId());
+        if (replicaToPromote == null) {
+            moveToUnassigned(failedPrimary, unassignedInfo);
+            for (ShardRouting unpromotableReplica : List.copyOf(assignedShards(failedPrimary.shardId()))) {
+                assert unpromotableReplica.primary() == false : unpromotableReplica;
+                assert unpromotableReplica.isPromotableToPrimary() == false : unpromotableReplica;
+                moveToUnassigned(
```

Comment:
I think we do rely on this invariant today in various places, although it's been an implicit assumption for such a long time that it's hard to know exactly where those places might be. It is verified by `IndexShardRoutingTable.Builder#noDuplicatePrimary` which was added pretty recently, but this assertion only didn't hold in a few rather specific unit tests. Mainly I think that's because of the `ReplicaAfterPrimaryActiveAllocationDecider` which would prevent the allocation of primary-free replicas in all realistic scenarios.

I'm pretty sure we can adjust things to cope with this requirement anyway, but it'll be a bit of an adventure and one best left for another day.